### PR TITLE
Give remote-test-client a configurable timeout

### PR DESCRIPTION
**Pull request by Hoverbear**
I'm doing some work with a remote host running tests using `x.py`'s remote test runner system.

After building the `remote-test-server` with:

```bash
./x build src/tools/remote-test-server --target aarch64-unknown-linux-gnu
```

I can transfer the `remote-test-server` to the remote and set up a port forwarded SSH connection, then I run:

```bash
TEST_DEVICE_ADDR=127.0.0.1:12345 ./x.py test library/core --target aarch64-unknown-linux-gnu
```

This works great if the host is nearby, however if the average round trip time is over 100ms (the timeout), it creates problems as it silently trips the timeout.

This PR checks an environment variable for a user configured timeout, defaulting to the old setting of 100 if it's not present. I picked the environment name because it seemed obvious, but there may be another option more suitable!

Also included is an explicit panic with a (hopefully) useful message if the deadline does get tripped. If it's more appropriate, I could make this a non-panic and simply a printed warning.

# Testing

To test, you could run the `remote-test-server` similar to as described in https://rustc-dev-guide.rust-lang.org/tests/running.html#running-tests-on-a-remote-machine.

```bash
./x build src/tools/remote-test-server --stage 2
./build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/remote-test-server -v --bind 127.0.0.1:12345
```

Then, in another terminal:

```bash
TCP_TIMEOUT=300 TEST_DEVICE_ADDR=127.0.0.1:12345 ./x.py test tests/ui
```

To trigger the timeout, you'll need to introduce latency into your loopback and tune the timeout down to something below that.

That message looks like this:

```
    Finished `release` profile [optimized] target(s) in 0.14s
Connecting to remote device 127.0.0.1:12345 ...
thread 'main' panicked at src/tools/remote-test-client/src/main.rs:91:72:
TCP timeout of 100ms exceeded, set `TCP_TIMEOUT` to a larger value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:07
```

### CODE REVIEW

File: src/tools/remote-test-client/src/main.rs [hunk 1]
```diff
@@ -69,16 +72,26 @@ fn spawn_emulator(target: &str, server: &Path, tmpdir: &Path, rootfs: Option<Pat
...
             t!(client.set_read_timeout(Some(dur)));
             t!(client.set_write_timeout(Some(dur)));
             if client.write_all(b"ping").is_ok() {
                 let mut b = [0; 4];
-                if client.read_exact(&mut b).is_ok() {
-                    break;
+                match client.read_exact(&mut b) {
+                    Ok(_) => break,
+                    Err(e) if e.kind() == io::ErrorKind::WouldBlock => panic!(
+                        "TCP timeout of {timeout}ms exceeded, set `{TCP_TIMEOUT_ENV}` to a larger value"
```
Comment (Mark-Simulacrum) [inline src/tools/remote-test-client/src/main.rs:92]: I'd have expected https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html#variant.TimedOut, not WouldBlock?

**Comment by Mark-Simulacrum** [comment]
I think I'd expect us to just increase the timeout rather than making it configurable. I'm not sure there's any significant value of 100ms vs. something like 2 seconds (which seems likely to be enough for your use case?)

File: src/tools/remote-test-client/src/main.rs (hunk 1 above)
Comment (Hoverbear) [reply src/tools/remote-test-client/src/main.rs:92]: That would be a more specific and appropriate error, yes! I wonder why I saw what I did while testing...

**Comment by Hoverbear** [comment]
I'm perfectly happy to simply increase the timeout to 2s! Let me do that instead!

**Comment by Hoverbear** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/127246

**Pull request closed without merging**